Implementation exclusion of files according to source file paths

ido-namely opened the pull request:
(including file globbing)
Goes to #56

ido-namely commented:
Would love to hear what you think is the best way to test this feature

MarcoRossignoli commented:
> Would love to hear what you think is the best way to test this feature

@tonerdo i'm not a "msbuild workflow" expert...but if we could have a simple guide/resouce link "how to test/debug" this project will grow much faster(very likely this is obvious for a msbuild workflow dev).

MarcoRossignoli commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -106,5 +108,26 @@ public static void DeleteHitsFile(string path)
+            return fileMatchResult.Files
+                .Select(
+                    f => System.IO.Path.GetFullPath(
+                        System.IO.Path.Combine(directoryInfo.ToString(), f.Path)
```
i think  instead of `directoryInfo.ToString()`  you can use `parentDir` right?
https://source.dot.net/#System.IO.FileSystem/System/IO/DirectoryInfo.cs,29
https://source.dot.net/#System.IO.FileSystem/System/IO/FileSystemInfo.cs,108

MarcoRossignoli replied:
nit: you can avoid `System.IO.` and use `Path.xxx`(using `System.IO;` )

MarcoRossignoli commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -3,12 +3,14 @@
 namespace Coverlet.Core.Helpers
 {
-    internal static class InstrumentationHelper
+    public static class InstrumentationHelper
```
@tonerdo maybe`[InternalsVisibleToAttribute]`?

tonerdo replied:
Please lets leave this as `internal`

MarcoRossignoli commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -106,5 +108,26 @@ public static void DeleteHitsFile(string path)
         }
+        
+        public static string[] GetExcludedFiles(string[] exclusionRules, string parentDir) {
+            if (exclusionRules == null || exclusionRules.Length == 0 ) return null;
```
IMHO usually i prefer `Array.Empty<string>()` https://docs.microsoft.com/en-us/dotnet/api/system.array.empty?view=netcore-2.0 less check condition for users.

MarcoRossignoli commented on src/coverlet.core/Helpers/InstrumentationHelper.cs:
```diff
@@ -3,12 +3,14 @@
 using System.IO;
 using System.Linq;
 using System.Reflection.PortableExecutable;
+using Microsoft.Extensions.FileSystemGlobbing;
```
i did't know this ext...thank's!

MarcoRossignoli commented:
@ido-namely one useful think i've learn from my contributions(on very first one) is to use "special key" to link PR to issue for simple "navigation".

https://help.github.com/articles/closing-issues-using-keywords/

if for instance on PR explanation you write:
```
closes #56  (including file globbing)
```

a link is automatically created on issue thread and you don't need to write anything on issue thread anymore but discussion move on PR.

abe545 commented on src/coverlet.core/Instrumentation/Instrumenter.cs:
```diff
@@ -8,19 +8,23 @@
         {
             _module = module;
             _identifier = identifier;
+            _excludedFiles = excludedFiles;
```
if you change this to `_excludedFiles = excludedFiles ?? Enumerable.Empty<string>()` you won't need the null checks elsewhere

MarcoRossignoli replied:
+1

abe545 commented on src/coverlet.core/Instrumentation/Instrumenter.cs:
```diff
@@ -61,6 +65,10 @@ private void InstrumentModule()
                     foreach (var method in type.Methods)
                     {
+                        var sourceFiles = method.DebugInformation.SequencePoints.Select(s => s.Document.Url).Distinct();
+                        if (_excludedFiles != null && sourceFiles.Any(_excludedFiles.Contains)) {
```
Odd that the method could be declared in multiple files (I guess this is the case with `partial methods` in `partial classes`). I'm wondering if in this case, you'd only want to exclude the method if `sourceFiles.All(...)`.  @tonerdo you're obviously a lot more familiar with this - am I interpreting this correctly?

abe545 replied:
They are part of the language: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/partial-method

If it isn't something that is supported in coverlet, then I guess we don't need to worry about it.

tonerdo replied:
OMG I've never even heard of `partial methods` until today. I guess we learn everyday. Taking my new found knowledge into consideration I'll still go with `.First(...)`, reason being that sequence points start at the body of the method and no matter how many partials of the same method you have, you can only have one implementation and the document that implementation is contained in is where all sequence points for that method will point to

tonerdo commented:
@ido-namely @MarcoRossignoli See steps to build and test project here: https://github.com/tonerdo/coverlet#building-the-project

ido-namely commented:
Hey @tonerdo , found some problems, please don't merge yet

MarcoRossignoli commented:
@tonerdo  maybe you already know this app https://github.com/apps/wip useful to permit to PR writer to block merge.

ido-namely commented:
@tonerdo I fixe the problem I faced.
I also added support for absolute paths, updated the readme file and the build.proj file to use the new flag.

I see there is some problem in appveyor, not sure if it's related to the project or not.

tonerdo commented on README.md:
```diff
@@ -78,13 +78,24 @@ The above command will automatically fail the build if the average code coverage
 Coverlet just uses the type name, so the attributes can be created under any namespace of your choosing.
 
+#### File Path  
+You can also ignore code coverage by specifing files using the `Exclude` property
```
Nit: You can also ignore specific source files from code coverage using the `Exclude` property

ryanmvt commented:
awesome! this looks great guys. Thank you @ido-namely for taking this on.
